File: composeApp/src/androidMain/kotlin/com/appointy/designsystem/components/DsIconButton.kt
package com.appointy.designsystem.components

import androidx.compose.foundation.layout.size
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.core.designsystem.tokens.components.button.ButtonColorRole
import com.core.designsystem.tokens.components.button.ButtonSize
import com.core.designsystem.tokens.components.iconbutton.IconButtonTokens

/**
 * DS Icon Button — MUI-style API.
 * Circular button for single icons, appropriate for app bars or toolbars.
 * Colors: primary, secondary, success, error, warning, info.
 * Sizes: small, medium, large.
 */
@Composable
fun DsIconButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    color: ButtonColorRole = ButtonColorRole.PRIMARY,
    size: ButtonSize = ButtonSize.MEDIUM,
    loading: Boolean = false,
    content: @Composable () -> Unit
) {
    val buttonSize = IconButtonTokens.getSize(size).dp
    val iconColor = Color(IconButtonTokens.getColor(color).toInt())

    IconButton(
        onClick = onClick,
        modifier = modifier.size(buttonSize),
        enabled = enabled && !loading,
        colors = IconButtonDefaults.iconButtonColors(contentColor = iconColor)
    ) {
        if (loading) {
            val loaderSize = when (size) {
                ButtonSize.SMALL -> 14.dp
                ButtonSize.MEDIUM -> 18.dp
                ButtonSize.LARGE -> 22.dp
            }
            CircularProgressIndicator(
                modifier = Modifier.size(loaderSize),
                color = iconColor,
                strokeWidth = 2.dp
            )
        } else {
            content()
        }
    }
}

File: iosApp/iosApp/DesignSystem/DsIconButton.swift
import SwiftUI
import Shared

/// DS Icon Button — MUI-style API.
/// Circular button for single icons, appropriate for app bars or toolbars.
/// Colors: primary, secondary, success, error, warning, info.
/// Sizes: small, medium, large.
struct DsIconButton<Content: View>: View {
    let action: () -> Void
    let content: Content
    var colorRole: ButtonColorRole
    var size: ButtonSize
    var disabled: Bool
    var loading: Bool

    init(
        action: @escaping () -> Void,
        color: ButtonColorRole = .primary,
        size: ButtonSize = .medium,
        disabled: Bool = false,
        loading: Bool = false,
        @ViewBuilder content: () -> Content
    ) {
        self.action = action
        self.content = content()
        self.colorRole = color
        self.size = size
        self.disabled = disabled
        self.loading = loading
    }

    var body: some View {
        let tokens = IconButtonTokens.shared
        let iconColor = Color(kmpArgb: tokens.getColor(colorRole: colorRole))
        let buttonSize = tokens.getSize(size: size)
        let iconSize = tokens.getIconSize(size: size)

        Button(action: action) {
            Group {
                if loading {
                    ProgressView()
                        .progressViewStyle(CircularProgressViewStyle(tint: iconColor))
                        .scaleEffect(buttonSize / 40.0) // Adjust spinner size based on button size
                } else {
                    content
                }
            }
            .font(.system(size: iconSize))
            .frame(width: buttonSize, height: buttonSize)
            .foregroundColor(iconColor)
            .contentShape(Circle())
            .opacity(disabled || loading ? 0.5 : 1)
        }
        .buttonStyle(.plain)
        .disabled(disabled || loading)
    }
}
